src/vodozemac/ecies.rs

```diff
@@ -0,0 +1,209 @@
+//! This module implements [ECIES](https://en.wikipedia.org/wiki/Integrated_Encryption_Scheme), the
+//! elliptic curve variant of the Integrated Encryption Scheme.
+//!
+//! Please take a look at the vodozemac documentation of this module for more
+//! info.
+
+#![allow(missing_debug_implementations)]
+use std::sync::{Arc, Mutex};
+
+use matrix_sdk_crypto::vodozemac::ecies;
+use wasm_bindgen::prelude::*;
+
+use super::Curve25519PublicKey;
+
+/// The result of an inbound ECIES channel establishment.
+#[wasm_bindgen(getter_with_clone)]
+pub struct InboundCreationResult {
+    /// The established ECIES channel.
+    pub channel: EstablishedEcies,
+    /// The plaintext of the initial message.
+    pub message: String,
+}
+
+/// The result of an outbound ECIES channel establishment.
+#[wasm_bindgen(getter_with_clone)]
+pub struct OutboundCreationResult {
+    /// The established ECIES channel.
+    pub channel: EstablishedEcies,
+    /// The initial encrypted message.
+    pub initial_message: String,
+}
+
+/// An unestablished ECIES session.
+#[wasm_bindgen]
+pub struct Ecies {
+    inner: Option<ecies::Ecies>,
+    public_key: Curve25519PublicKey,
+}
+
+#[wasm_bindgen]
+impl Ecies {
+    /// Create a new, random, unestablished ECIES session.
+    ///
+    /// This method will use the `MATRIX_QR_CODE_LOGIN` info.
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Self {
+        let inner = ecies::Ecies::new();
+        let public_key = inner.public_key().into();
+
+        Self { inner: Some(inner), public_key }
+    }
+
+    /// Get our [`Curve25519PublicKey`].
+    ///
+    /// This public key needs to be sent to the other side to be able to
+    /// establish an ECIES channel.
+    pub fn public_key(&self) -> Curve25519PublicKey {
+        self.public_key.clone()
+    }
+
+    fn used_up_error() -> JsError {
+        JsError::new("The ECIES channel was already established and used up.")
+    }
+
+    /// Create a [`EstablishedEcies`] from an initial message encrypted by the
+    /// other side.
+    pub fn establish_inbound_channel(
+        &mut self,
+        initial_message: &str,
+    ) -> Result<InboundCreationResult, JsError> {
+        let message = ecies::InitialMessage::decode(&initial_message)?;
+        let result = self
+            .inner
+            .take()
+            .ok_or_else(Self::used_up_error)?
+            .establish_inbound_channel(&message)?;
```

Comment:
If the answer to the above question is "no", then: is it worth putting the `self.inner.take()` back into the `self.inner`, if this line yields an error result?

Reply:
No, the channel is supposed to be ephemeral for security reasons, and if things fail you shouldn't try to reuse the generated keys.